### ISSUE TITLE
Google WebView warning on Android 4.0.3 emulator

### ISSUE THREAD

**Issue by dg76** [openid/AppAuth-Android#210]
It works fine now for the "Samsung Galaxy Nexus with Android 4.3" #196 but the warning is still displayed on an Android emulator with the "Android 4.0.3 (with Google APIs)" system image. Can this be fixed, too?

Thank you very much!

Kind regards

Dominique

**Comment by iainmcgin** [comment]
AppAuth requires API level 16 or higher, so you should not be able to run the app on a 4.0.3 image. Also, I was unable to reproduce with an API level 16 emulator.

**Comment by dg76** [comment]
I understand that the library is not supposed to run in API level 15. But this isn't a problem of the library, it is a problem of Google's login website, which does not recognize that a browser instead of a webview is used.

I have tried to run it on API level 15 in case users contact me after April 20th and are looking for a solution for older devices. If there isn't an official library for older devices I would try to find a solution myself but this won't work if the Google login website itself does not support older devices any more.

I have reported it here because here seem to be people (like you) who know how to contact the Google team that manages the whitelist for the Google login website. If there would be a project just for that website, I would have reported it there, because this problem is not caused by the library but by the Google login website.

Does that mean that the Google login website does not support Android 4.0.3 and older devices any more, regardless which library or method is used to open the website?

Here is a screenshot of the problem:

![screen_shot_2017-04-01_at_09_00_11](https://cloud.githubusercontent.com/assets/17218461/24576298/d1431270-16b9-11e7-8a52-8742b65de6fe.png)

I am sorry that this message became so long. Thank you very much for reading it!

Kind regards

Dominique

**Comment by iainmcgin** [comment]
When I get a chance I'll see if I can reproduce the problem with an API 15 emulator.

Side question: Which API level would you be looking to support? I have discussed lowering the version requirement of the custom tabs library to API 9 (Gingerbread) with the custom tab engineers. This is the only dependency of AppAuth at present, so if I had justification for lowering the minimum API version of that library, I could potentially lower the minimum API version of AppAuth too.

However, I believe in practice that API 16 is a reasonable lower bound for our library right now; 16+ covers [98% of all Android devices](https://developer.android.com/about/dashboards/index.html) that Google can see.

**Comment by dg76** [comment]
For a long time I have supported API level 4 (while nevertheless at the same time supporting features from the newest API levels). A few months ago, in December 2016, I have changed the minimum API level to 9. Then a server URL changed that was used by a small percentage of the users. I updated the app with API level 9. But then a user with API level 8 contacted me and also needed a new version (which I created by creating a special branch for such older API levels). That means there are still people out there using these old versions. 

So while I am not sure how many users are affected it would be great if API level 9 could be supported.

In the future I will probably also use a higher API level for the main version of my app (due to multidex problems on lower API levels) but for existing users I would like to keep the app at least fully functional even on lower API levels. I will (and already do) use Google Play's feature to provide different APKs for different API levels.

So if you could make the AppAuth library compatible to API level 9 that would be great! 👍

Thank you very much!

Kind regards

Dominique

**Comment by iainmcgin** [comment]
Upon consideration we're not willing to take on the support burden of devices prior to API 16, so we won't be changing our current minimum.

Could you provide me the package name and version number of the system browser on your emulator?

**Comment by dg76** [comment]
Ok, I will see if I will get any requests from users using an older version and will then perhaps try to make it compatible with older versions myself.

I have tried it myself with the emulator but the warning does not appear any more. It looks as if it was already fixed. Thank you very much! Nevertheless I have checked the package name and the version. The package name is "com.android.browser" and the version is "4.0.4-3093080".

Kind regards

Dominique
